instance/models/utils.py

```diff
@@ -44,3 +55,263 @@ def save(self, *args, **kwargs):
+
+
+class ClassProperty(property):
+    """ Same as built-in 'property' global but also works when accessed as a class attribute """
+    def __get__(self, cls, owner):
+        return self.fget.__get__(None, owner)()  # pylint: disable=no-member
+
+
+class ResourceState:
+    """
+    A [finite state machine] state class representing the overall state of a resource,
+    such as a server, a database, a DNS entry, etc.
+    """
+    # state_id: A string uniquely identifying this state
+    state_id = None
+
+    @ClassProperty
+    @classmethod
+    def name(cls):
+        """
+        name: Human-readable name of this state, suitable for display as "Status: _______"
+
+        Doesn't need to be a property; override in the subclass using just 'name = "Name"'
+        """
+        return cls.__name__
+
+    @ClassProperty
+    @classmethod
+    def description(cls):
+        """
+        description: Human-readable explanation of this state (1-2 sentences)
+
+        Defaults to the state class's docstring, but you can override it.
+        Doesn't need to be a property; override in the subclass using just 'description = "..."'
+        """
+        return cls.__doc__.strip()
+
+    # Constants:
+    PROGRESS_UNKNOWN = -1
+
+    def __str__(self):
+        return "{} [{}]".format(self.name, self.state_id)
+
+    def __eq__(self, obj):
+        """
+        Syntactic sugar to make comparing states easier.
+
+        obj should be a ResourceState class or instance.
+        """
+        try:
+            return issubclass(self.__class__, obj)
+        except TypeError:  # The provided argument is not a class
+            return type(obj) is type(self)
+
+    @classmethod
+    def one_of(cls, *state_classes):
+        """ Syntactic sugar to make comparing an instance to a set of types easier """
+        return issubclass(cls, state_classes)
+
+    def __init__(self, resource, state_manager):
+        """ Instantiate this state, saving a reference to the parent resource """
+        self._resource = resource
+        self._state_manager = state_manager
+
+    class Enum:
+        """
+        Syntacic sugar for declaring a group of ResourceState classes inside a class.
+
+        Use like:
+
+        class MyStates(ResourceState.Enum):
+            class State1(ResourceState):
+                state_id = '1'
+            class State2(ResourceState):
+                state_id = '2'
+
+        or if states are already declared, use like:
+        class MyStates(ResourceState.Enum):
+            State1, State2 = State1, State2
+
+        Then you can use MyStates.states to get a list of the state classes.
+        """
+        @ClassProperty
+        @classmethod
+        def states(cls):
+            """
+            Get an iterable listing all the classes defined within this class
+            """
+            return tuple(state for name, state in cls.__dict__.items() if name[:1] != '_' and name != 'states')
```

Comment:
This approach won't work together with inheritance, since `cls.__dict__` (or equivalently, `vars(cls)`) does not contain inherited members.  I suggest using `inspect.getmembers()` instead.  (Filtering out `states` isn't necessary with the current approach, since it won't be in the list, but will be necessary when using `inspect.getmembers()`.)